File: .lib/BRXSDK_Bcad_V23_2_03/samples/VbBrxMgd/EditorSample.vb
' system 
Imports System.Text
Imports System.Runtime.InteropServices
Imports System.IO


' ODA
Imports Teigha.Runtime
Imports Teigha.DatabaseServices
Imports Teigha.Geometry

' Bricsys
Imports Bricscad.ApplicationServices
Imports Bricscad.Runtime
Imports Bricscad.EditorInput

' com
Imports BricscadDb
Imports BricscadApp

' alias
Imports _AcRx = Teigha.Runtime
Imports _AcAp = Bricscad.ApplicationServices
Imports _AcDb = Teigha.DatabaseServices
Imports _AcGe = Teigha.Geometry
Imports _AcEd = Bricscad.EditorInput
Imports _AcGi = Teigha.GraphicsInterface
Imports _AcClr = Teigha.Colors
Imports _AcWnd = Bricscad.Windows

Namespace CsBrxMgd
    Friend Class EditorSample
        ' Methods
        Public Shared Sub do_getDistance()
            Dim editor As Editor = Application.DocumentManager.MdiActiveDocument.Editor
            Dim ppo As New PromptPointOptions("Select Point")
            Dim ppr As PromptPointResult = editor.GetPoint(ppo)
            If (ppr.Status <> PromptStatus.OK) Then
                Throw New System.Exception("GetPoint failed")
            End If
            Dim pdo As New PromptDistanceOptions("Select next Point")
            pdo.UseBasePoint = True
            pdo.BasePoint = ppr.Value
            pdo.UseDashedLine = True
            Dim pdr As PromptDoubleResult = editor.GetDistance(pdo)
            If (pdr.Status <> PromptStatus.OK) Then
                Throw New System.Exception("GetDistance failed")
            End If
            Application.ShowAlertDialog(pdr.ToString)
        End Sub

        Public Shared Sub do_getEntity()
            Dim editor As Editor = Application.DocumentManager.MdiActiveDocument.Editor
            Dim peo As New PromptEntityOptions("Select a Hatch")
            peo.SetRejectMessage("Not a Hatch!")
            peo.AddAllowedClass(GetType(Hatch), True)
            Dim per As PromptEntityResult = editor.GetEntity(peo)
            If (per.Status = PromptStatus.OK) Then
                Using action As Transaction = HostApplicationServices.WorkingDatabase.TransactionManager.StartTransaction
                    Dim hatch As Hatch = TryCast(action.GetObject(per.ObjectId, OpenMode.ForWrite), Hatch)
                    If (Not hatch Is Nothing) Then
                        Application.ShowAlertDialog(("Area is " & hatch.Area.ToString))
                    End If
                    action.Commit()
                End Using
            End If
        End Sub

        Public Shared Sub do_getEntity2()
            Dim editor As Editor = Application.DocumentManager.MdiActiveDocument.Editor
            Dim peo As New PromptEntityOptions("Select a Line")
            peo.SetRejectMessage("Not a Line!")
            peo.AddAllowedClass(GetType(Line), True)
            Dim per As PromptEntityResult = editor.GetEntity(peo)
            If (per.Status = PromptStatus.OK) Then
                Using action As Transaction = HostApplicationServices.WorkingDatabase.TransactionManager.StartTransaction
                    Dim line As Line = TryCast(action.GetObject(per.ObjectId, OpenMode.ForWrite), Line)
                    If (Not line Is Nothing) Then
                        Application.ShowAlertDialog(("Length is " & line.Length.ToString))
                    End If
                    action.Commit()
                End Using
            End If
        End Sub

        Public Shared Sub do_getKeyWords()
            Dim editor As Editor = Application.DocumentManager.MdiActiveDocument.Editor
            Dim pop As New PromptKeywordOptions("this is a test")
            pop.AppendKeywordsToMessage = True
            pop.AllowNone = False
            pop.Keywords.Add("A_Kwd")
            pop.Keywords.Add("B_Kwd")
            pop.Keywords.Add("C_Kwd")
            Application.ShowAlertDialog(editor.GetKeywords(pop).ToString)
        End Sub

        Public Shared Sub do_getPoint()
            Dim editor As Editor = Application.DocumentManager.MdiActiveDocument.Editor
            Dim pop As New PromptPointOptions("Select Point")
            Application.ShowAlertDialog(editor.GetPoint(pop).ToString)
        End Sub

        Public Shared Sub do_getSelectionSet()
            Dim editor As Editor = Application.DocumentManager.MdiActiveDocument.Editor
            Dim dxfs As TypedValue() = New TypedValue() {New TypedValue(0, "POINT")}
            Dim sfilter As New SelectionFilter(dxfs)
            Dim pso As New PromptSelectionOptions
            pso.MessageForAdding = "Select Points"
            pso.MessageForRemoval = "Remove Points"
            pso.AllowDuplicates = False
            Dim res As PromptSelectionResult = editor.GetSelection(pso, sfilter)
            If (res.Status = PromptStatus.OK) Then
                Dim sset As SelectionSet = res.Value
                editor.WriteMessage(vbLf & "Found {0} points", New Object() {sset.Count})
                editor.WriteMessage(vbLf & "{0}", New Object() {EditorSample.getPointLocations(sset.GetObjectIds).Count})
            End If
        End Sub

        Public Shared Sub do_getString()
            Dim editor As Editor = Application.DocumentManager.MdiActiveDocument.Editor
            Dim pop As New PromptStringOptions("this is a test")
            Application.ShowAlertDialog(editor.GetString(pop).ToString)
        End Sub

        Public Shared Sub do_MainWindow()
            Dim editor As Editor = Application.DocumentManager.MdiActiveDocument.Editor
            Dim w As _AcWnd.Window = _AcAp.Application.MainWindow
            w.Text = "*!*BricsCAD Rocks*!*"
            w.WindowState = FormWindowState.Minimized
            If (w.WindowState = FormWindowState.Minimized) Then
                editor.WriteMessage("Minimized")
            End If
            w.WindowState = FormWindowState.Normal
            If (w.WindowState = FormWindowState.Normal) Then
                editor.WriteMessage("Normal")
            End If
            w.WindowState = FormWindowState.Maximized
            If (w.WindowState = FormWindowState.Maximized) Then
                editor.WriteMessage("Maximized")
            End If
        End Sub

        Public Shared Function getPointLocations(ByVal ids As ObjectId()) As Point3dCollection
            Dim locations As New Point3dCollection
            Using action As Transaction = HostApplicationServices.WorkingDatabase.TransactionManager.StartTransaction
                Dim id As ObjectId
                For Each id In ids
                    Dim point As DBPoint = TryCast(action.GetObject(id, OpenMode.ForRead), DBPoint)
                    If (Not point Is Nothing) Then
                        locations.Add(point.Position)
                    End If
                Next
                action.Commit()
            End Using
            Return locations
        End Function

        Public Shared Sub OnPointMonitor(ByVal sender As Object, ByVal eventArgs As PointMonitorEventArgs)
            Dim null As ObjectId = ObjectId.Null
            Dim editor As Editor = DirectCast(sender, Editor)
            Dim document As Document = editor.Document
            If editor.IsQuiescent Then
                Dim context As InputPointContext = eventArgs.Context
                context.DrawContext.Geometry.Circle(context.RawPoint, 10, Vector3d.ZAxis)
                Dim names As String = Nothing
                Dim paths As FullSubentityPath() = eventArgs.Context.GetPickedEntities
                Using transaction As Transaction = document.TransactionManager.StartTransaction
                    Array.Sort(Of FullSubentityPath)(paths, New Comparison(Of FullSubentityPath)(AddressOf EditorSample.sortByIds))
                    Dim path As FullSubentityPath
                    For Each path In paths
                        Dim ids As ObjectId() = path.GetObjectIds
                        If (ids.Length > 0) Then
                            Dim bref As BlockReference = TryCast(transaction.GetObject(ids(ids.GetLowerBound(0)), OpenMode.ForRead), BlockReference)
                            If (Not bref Is Nothing) Then
                                names = (names & String.Format(vbLf & "Block Name= {0} {1}", bref.Name, path.SubentId.IndexPtr.ToInt32))
                                Exit For
                            End If
                        End If
                    Next
                    transaction.Commit()
                End Using
                eventArgs.AppendToolTipText(names)
            End If
        End Sub

        Public Shared Function sortByIds(ByVal a As FullSubentityPath, ByVal b As FullSubentityPath) As Integer
            Return a.SubentId.IndexPtr.ToInt32.CompareTo(b.SubentId.IndexPtr.ToInt32)
        End Function

    End Class
End Namespace

File: autoframecad/libs/brx/samples/CsBrxMgd/CsBrxMgd/EditorSample.cs
// Property of Bricsys NV. All rights reserved.
// This file is part of the BRX SDK, and its use is subject to the terms
// of the BRX SDK license agreement.
/////////////////////////////////////////////////////////////////////////

// system 
using System;
using System.Text;
using System.Diagnostics;
using System.Collections.Generic;
using System.Runtime.InteropServices;
using System.IO;

// alias
using _AcAp = Bricscad.ApplicationServices;
using _AcDb = Teigha.DatabaseServices;
using _AcGe = Teigha.Geometry;
using _AcEd = Bricscad.EditorInput;

namespace CsBrxMgd
{
    class EditorSample
    {
        //samp25
        public static void do_getSelectionSet()
        {
            _AcEd.Editor editor = _AcAp.Application.DocumentManager.MdiActiveDocument.Editor;
            _AcDb.TypedValue[] dxfs = new _AcDb.TypedValue[] { new _AcDb.TypedValue((int)_AcDb.DxfCode.Start, "POINT") };
            _AcEd.SelectionFilter sfilter = new _AcEd.SelectionFilter(dxfs);

            _AcEd.PromptSelectionOptions pso = new _AcEd.PromptSelectionOptions();
            pso.MessageForAdding = "Select Points";
            pso.MessageForRemoval = "Remove Points";
            pso.AllowDuplicates = false;

            _AcEd.PromptSelectionResult res = editor.GetSelection(pso, sfilter);
            if (res.Status == _AcEd.PromptStatus.OK)
            {
                _AcEd.SelectionSet sset = res.Value;

                editor.WriteMessage("\nFound {0} points", sset.Count);
                editor.WriteMessage("\n{0}", getPointLocations(sset.GetObjectIds()).Count);
            }
        }

        public static _AcGe.Point3dCollection getPointLocations(_AcDb.ObjectId[] ids)
        {
            _AcGe.Point3dCollection locations = new _AcGe.Point3dCollection();
            _AcDb.Database database = _AcDb.HostApplicationServices.WorkingDatabase;
            _AcDb.TransactionManager manager = database.TransactionManager;
            using (_AcDb.Transaction action = manager.StartTransaction())
            {
                foreach (_AcDb.ObjectId id in ids)
                {
                    _AcDb.DBPoint point = action.GetObject(id, _AcDb.OpenMode.ForRead) as _AcDb.DBPoint;
                    if (point != null)
                        locations.Add(point.Position);
                }
                action.Commit();
            }
            return locations;
        }

        //samp18
        public static void do_getString()
        {
            _AcEd.Editor editor = _AcAp.Application.DocumentManager.MdiActiveDocument.Editor;
            _AcEd.PromptStringOptions pop = new _AcEd.PromptStringOptions("this is a test");
            _AcEd.PromptResult res = editor.GetString(pop);
            _AcAp.Application.ShowAlertDialog(res.ToString());
        }

        public static void do_getKeyWords()
        {
            _AcEd.Editor editor = _AcAp.Application.DocumentManager.MdiActiveDocument.Editor;
            _AcEd.PromptKeywordOptions pop = new _AcEd.PromptKeywordOptions("this is a test");
            pop.AppendKeywordsToMessage = true;
            pop.AllowNone = false;
            pop.Keywords.Add("A_Kwd");
            pop.Keywords.Add("B_Kwd");
            pop.Keywords.Add("C_Kwd");
            _AcEd.PromptResult res = editor.GetKeywords(pop);
            _AcAp.Application.ShowAlertDialog(res.ToString());
        }

        public static void do_getPoint()
        {
            _AcEd.Editor editor = _AcAp.Application.DocumentManager.MdiActiveDocument.Editor;
            _AcEd.PromptPointOptions pop = new _AcEd.PromptPointOptions("Select Point");
            _AcEd.PromptPointResult res = editor.GetPoint(pop);
            _AcAp.Application.ShowAlertDialog(res.ToString());
        }

        public static void do_getDistance()
        {
            _AcEd.Editor editor = _AcAp.Application.DocumentManager.MdiActiveDocument.Editor;
            _AcEd.PromptPointOptions ppo = new _AcEd.PromptPointOptions("Select Point");
            _AcEd.PromptPointResult ppr = editor.GetPoint(ppo);

            if (ppr.Status != _AcEd.PromptStatus.OK)
                throw new System.Exception("GetPoint failed");

            _AcEd.PromptDistanceOptions pdo = new _AcEd.PromptDistanceOptions("Select next Point");

            pdo.AllowNone = true;
            pdo.UseBasePoint = true;
            pdo.BasePoint = ppr.Value;
            pdo.UseDashedLine = true;

            _AcEd.PromptDoubleResult pdr = editor.GetDistance(pdo);

            if (pdr.Status != _AcEd.PromptStatus.OK)
                throw new System.Exception("GetDistance failed");
            _AcAp.Application.ShowAlertDialog(pdr.ToString());
        }

        // command samp22
        public static void do_getEntity()
        {
            _AcEd.Editor editor = _AcAp.Application.DocumentManager.MdiActiveDocument.Editor;
            _AcEd.PromptEntityOptions peo = new _AcEd.PromptEntityOptions("Select a Hatch");
            peo.SetRejectMessage("Not a Hatch!");
            peo.AddAllowedClass(typeof(_AcDb.Hatch), true);
            _AcEd.PromptEntityResult per = editor.GetEntity(peo);
            if (per.Status == _AcEd.PromptStatus.OK)
            {
                _AcDb.Database database = _AcDb.HostApplicationServices.WorkingDatabase;
                _AcDb.TransactionManager manager = database.TransactionManager;
                using (_AcDb.Transaction action = manager.StartTransaction())
                {
                    _AcDb.Hatch hatch = action.GetObject(per.ObjectId, _AcDb.OpenMode.ForWrite) as _AcDb.Hatch;
                    if (hatch != null)
                        _AcAp.Application.ShowAlertDialog("Area is " + hatch.Area.ToString());
                    action.Commit();
                }
            }
        }

        // command samp23
        public static void do_getEntity2()
        {
            _AcEd.Editor editor = _AcAp.Application.DocumentManager.MdiActiveDocument.Editor;
            _AcEd.PromptEntityOptions peo = new _AcEd.PromptEntityOptions("Select a Line");
            peo.SetRejectMessage("Not a Line!");
            peo.AddAllowedClass(typeof(_AcDb.Line), true);
            _AcEd.PromptEntityResult per = editor.GetEntity(peo);
            if (per.Status == _AcEd.PromptStatus.OK)
            {
                _AcDb.Database database = _AcDb.HostApplicationServices.WorkingDatabase;
                _AcDb.TransactionManager manager = database.TransactionManager;
                using (_AcDb.Transaction action = manager.StartTransaction())
                {
                    _AcDb.Line line = action.GetObject(per.ObjectId, _AcDb.OpenMode.ForWrite) as _AcDb.Line;
                    if (line != null)
                        _AcAp.Application.ShowAlertDialog("Length is " + line.Length.ToString());
                    action.Commit();
                }
            }
        }

        public static void do_MainWindow()
        {
            _AcEd.Editor editor = _AcAp.Application.DocumentManager.MdiActiveDocument.Editor;
            //AcAp.Application.MainWindow gives us access to the main window
            //so we can access/change its properties
            Bricscad.Windows.Window w = _AcAp.Application.MainWindow;
            w.Text = "*!*BricsCAD Rocks*!*";

            w.WindowState = System.Windows.Forms.FormWindowState.Minimized;
            if (w.WindowState == System.Windows.Forms.FormWindowState.Minimized)
                editor.WriteMessage("Minimized");

            w.WindowState = System.Windows.Forms.FormWindowState.Normal;
            if (w.WindowState == System.Windows.Forms.FormWindowState.Normal)
                editor.WriteMessage("Normal");

            w.WindowState = System.Windows.Forms.FormWindowState.Maximized;
            if (w.WindowState == System.Windows.Forms.FormWindowState.Maximized)
                editor.WriteMessage("Maximized");
        }

        public static int sortByIds(_AcDb.FullSubentityPath a, _AcDb.FullSubentityPath b)
        {
            return a.SubentId.IndexPtr.ToInt32().CompareTo(b.SubentId.IndexPtr.ToInt32());
        }

        public static void OnPointMonitor(object sender, _AcEd.PointMonitorEventArgs eventArgs)
        {
            _AcDb.ObjectId id = _AcDb.ObjectId.Null;
            _AcEd.Editor editor = (_AcEd.Editor)sender;
            _AcAp.Document document = editor.Document;
            if (!editor.IsQuiescent)
                return;
            _AcEd.InputPointContext context = eventArgs.Context;
            Teigha.GraphicsInterface.ViewportDraw draw = context.DrawContext;
            draw.Geometry.Circle(context.RawPoint, 10, _AcGe.Vector3d.ZAxis);
            string names = null;
            _AcDb.FullSubentityPath[] paths = eventArgs.Context.GetPickedEntities();
            using (_AcDb.Transaction transaction = document.TransactionManager.StartTransaction())
            {
                Array.Sort(paths, sortByIds);
                foreach (_AcDb.FullSubentityPath path in paths)
                {
                    _AcDb.ObjectId[] ids = path.GetObjectIds();
                    if (ids.Length > 0)
                    {
                        _AcDb.BlockReference bref = transaction.GetObject(
                            ids[ids.GetLowerBound(0)], _AcDb.OpenMode.ForRead) as _AcDb.BlockReference;
                        if (bref != null)
                        {
                            names += string.Format("\nBlock Name= {0} {1}", bref.Name, path.SubentId.IndexPtr);
                            break;
                        }
                    }
                }
                transaction.Commit();
            }
            eventArgs.AppendToolTipText(names);
        }

    }
}
